File: src/bmsx_cpp/machine/specs.cpp
#include "machine/specs.h"

#include "machine/runtime/timing.h"
#include "rompack/assets.h"

#include <stdexcept>

namespace bmsx {

static constexpr const char* kCpuHzMissingMessage = "[RuntimeMachineSpecs] machine.specs.cpu.cpu_freq_hz is required.";
static constexpr const char* kCpuHzInvalidMessage = "[RuntimeMachineSpecs] machine.specs.cpu.cpu_freq_hz must be a positive integer.";
static constexpr const char* kUfpsMissingMessage = "[RuntimeMachineSpecs] machine.ufps is required.";
static constexpr const char* kUfpsInvalidMessage = "[RuntimeMachineSpecs] machine.ufps must be greater than 1 Hz.";

template<typename Predicate>
static bool tryResolveValue(const std::optional<i64>& value, i64& out, Predicate predicate) {
	if (!value || !predicate(*value)) {
		return false;
	}
	out = *value;
	return true;
}

i64 requirePositiveManifestValue(const std::optional<i64>& value, const char* missingMessage, const char* invalidMessage) {
	if (!value) {
		throw std::runtime_error(missingMessage);
	}
	if (*value <= 0) {
		throw std::runtime_error(invalidMessage);
	}
	return *value;
}

// start value-or-boundary -- manifest defaults are resolved at the manifest boundary.
i64 resolvePositiveManifestValue(const std::optional<i64>& value, i64 defaultValue, const char* invalidMessage) {
	const i64 resolved = value.value_or(defaultValue);
	if (resolved <= 0) {
		throw std::runtime_error(invalidMessage);
	}
	return resolved;
}
// end value-or-boundary

i64 requireManifestValueAbove(const std::optional<i64>& value, i64 minimumExclusive, const char* missingMessage, const char* invalidMessage) {
	if (!value) {
		throw std::runtime_error(missingMessage);
	}
	if (*value <= minimumExclusive) {
		throw std::runtime_error(invalidMessage);
	}
	return *value;
}

bool tryResolveCpuHz(const MachineManifest& manifest, i64& outHz) {
	return tryResolveValue(manifest.cpuHz, outHz, [](i64 hz) {
		return hz > 0;
	});
}

i64 resolveCpuHz(const MachineManifest& manifest) {
	return requirePositiveManifestValue(manifest.cpuHz, kCpuHzMissingMessage, kCpuHzInvalidMessage);
}

bool tryResolveUfpsScaled(const MachineManifest& manifest, i64& outUfpsScaled) {
	return tryResolveValue(manifest.ufpsScaled, outUfpsScaled, [](i64 ufpsScaled) {
		return ufpsScaled > HZ_SCALE;
	});
}

i64 resolveUfpsScaled(const MachineManifest& manifest) {
	return requireManifestValueAbove(manifest.ufpsScaled, HZ_SCALE, kUfpsMissingMessage, kUfpsInvalidMessage);
}

} // namespace bmsx

File: machine/ts/machine/specs.ts
import { getMachinePerfSpecs, type MachineManifest, type Viewport } from '../rompack/format';
import { HZ_SCALE } from './runtime/timing/constants';

const CPU_HZ_MISSING_MESSAGE = '[RuntimeMachineSpecs] machine.specs.cpu.cpu_freq_hz is required.';
const CPU_HZ_INVALID_MESSAGE = '[RuntimeMachineSpecs] machine.specs.cpu.cpu_freq_hz must be a positive integer.';
const UFPS_MISSING_MESSAGE = '[RuntimeMachineSpecs] machine.ufps is required.';
const UFPS_INVALID_MESSAGE = '[RuntimeMachineSpecs] machine.ufps must be greater than 1 Hz.';

export type ResolvedManifestNumber = { value: number };

type ManifestNumberPredicate = (value: number) => boolean;

function resolveValue(value: number | undefined, out: ResolvedManifestNumber, predicate: ManifestNumberPredicate): boolean {
	if (value === undefined || !predicate(value)) {
		return false;
	}
	out.value = value;
	return true;
}

export function requirePositiveManifestValue(value: number | undefined, missingMessage: string, invalidMessage: string): number {
	if (value === undefined) {
		throw new Error(missingMessage);
	}
	if (!Number.isSafeInteger(value) || value <= 0) {
		throw new Error(invalidMessage);
	}
	return value;
}

export function resolvePositiveManifestValue(value: number | undefined, defaultValue: number, invalidMessage: string): number {
	const resolved = value === undefined ? defaultValue : value;
	if (!Number.isSafeInteger(resolved) || resolved <= 0) {
		throw new Error(invalidMessage);
	}
	return resolved;
}

export function requireManifestValueAbove(value: number | undefined, minimumExclusive: number, missingMessage: string, invalidMessage: string): number {
	if (value === undefined) {
		throw new Error(missingMessage);
	}
	if (!Number.isSafeInteger(value) || value <= minimumExclusive) {
		throw new Error(invalidMessage);
	}
	return value;
}

export function resolveCpuHz(machine: MachineManifest, outHz: ResolvedManifestNumber): boolean;
export function resolveCpuHz(machine: MachineManifest): number;
export function resolveCpuHz(machine: MachineManifest, outHz?: ResolvedManifestNumber): number | boolean {
	const value = getMachinePerfSpecs(machine).cpu_freq_hz;
	if (outHz !== undefined) {
		return resolveValue(value, outHz, hz => Number.isSafeInteger(hz) && hz > 0);
	}
	return requirePositiveManifestValue(value, CPU_HZ_MISSING_MESSAGE, CPU_HZ_INVALID_MESSAGE);
}

export function resolveUfpsScaled(machine: MachineManifest, outUfpsScaled: ResolvedManifestNumber): boolean;
export function resolveUfpsScaled(machine: MachineManifest): number;
export function resolveUfpsScaled(machine: MachineManifest, outUfpsScaled?: ResolvedManifestNumber): number | boolean {
	const value = getMachinePerfSpecs(machine).ufps;
	if (outUfpsScaled !== undefined) {
		return resolveValue(value, outUfpsScaled, ufpsScaled => Number.isSafeInteger(ufpsScaled) && ufpsScaled > HZ_SCALE);
	}
	return requireManifestValueAbove(value, HZ_SCALE, UFPS_MISSING_MESSAGE, UFPS_INVALID_MESSAGE);
}

function runtimeSpecFault(message: string): Error {
	return new Error(`Runtime fault: ${message}`);
}

export function resolvePositiveSafeInteger(value: number | undefined, label: string): number {
	if (value === undefined) {
		throw runtimeSpecFault(`${label} is required.`);
	}
	if (!Number.isSafeInteger(value) || value <= 0) {
		throw runtimeSpecFault(`${label} must be a positive safe integer.`);
	}
	return value;
}

export function resolveRuntimeRenderSize(machine: { render_size: { width: number; height: number; } }): Viewport {
	const width = resolvePositiveSafeInteger(machine.render_size.width, 'machine.render_size.width');
	const height = resolvePositiveSafeInteger(machine.render_size.height, 'machine.render_size.height');
	return { width, height };
}
